forms-flow-web/src/services/UserService.js

```diff
@@ -0,0 +1,157 @@
+ /* istanbul ignore file */
+import {
+  ROLES,
+  USER_RESOURCE_FORM_ID,
+  Keycloak_Client,
+  ANONYMOUS_USER,
+  ANONYMOUS_ID,
+  FORMIO_JWT_SECRET
+} from "../constants/constants";
+import {
+  setUserRole,
+  setUserToken,
+  setUserDetails,
+} from "../actions/bpmActions";
+import {BPM_BASE_URL} from "../apiManager/endpoints/config";
+import {AppConfig} from '../config';
+import {WEB_BASE_URL , WEB_BASE_CUSTOM_URL} from "../apiManager/endpoints/config";
+
+import {_kc} from "../constants/tenantConstant";
+
+const jwt = require("jsonwebtoken");
+
+/**
+ * Initializes Keycloak instance and calls the provided callback function if successfully authenticated.
+ *
+ * @param onAuthenticatedCallback
+ */
+// const KeycloakData = new Keycloak(tenantDetail);
+
+
+const initKeycloak = (store, ...rest) => {
+  const done = rest.length ? rest[0] : () => {};
+  KeycloakData
+    .init({
+      onLoad: "check-sso",
+      promiseType: "native",
+      silentCheckSsoRedirectUri:
+        window.location.origin + "/silent-check-sso.html",
+      pkceMethod: "S256",
+      checkLoginIframe: false
+    })
+    .then((authenticated) => {
+      if (authenticated) {
+        if (KeycloakData.resourceAccess[Keycloak_Client]) {
+          const UserRoles = KeycloakData.resourceAccess[Keycloak_Client].roles;
+          store.dispatch(setUserRole(UserRoles));
+          store.dispatch(setUserToken(KeycloakData.token));
+          //Set Cammunda/Formio Base URL
+          setApiBaseUrlToLocalStorage();
+
+          let roles = [];
+          for (let i = 0; i < UserRoles.length; i++) {
+            const roleData = ROLES.find((x) => x.title === UserRoles[i]);
+            if (roleData) {
+              roles = roles.concat(roleData.id);
+            }
+          }
+          KeycloakData.loadUserInfo().then((res) => store.dispatch(setUserDetails(res)));
+          const email = KeycloakData.tokenParsed.email || "external";
+          authenticateFormio(email, roles);
+          // onAuthenticatedCallback();
+          done(null, KeycloakData);
+          refreshToken(store);
+        } else {
+          doLogout();
+        }
+      } else {
+        console.warn("not authenticated!");
+        doLogin();
+      }
+    });
+};
+let refreshInterval;
+const refreshToken = (store) => {
+  refreshInterval = setInterval(() => {
+    KeycloakData && KeycloakData.updateToken(-1).then((refreshed)=> {
```

Comment:
If I remember correctly, this one was `updateToken(5)` instead of `-1` 🙃

Reply:
Just so we refresh the token when it has < 5 min left until expiry

Reply:
Oops! You're right. let me fix it.